File: server/src/models/crag.py
from sqlalchemy import func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.ext.hybrid import hybrid_property

from extensions import db
from models.area import Area
from models.ascent import Ascent
from models.base_entity import BaseEntity
from models.enums.searchable_item_type_enum import SearchableItemTypeEnum
from models.line import Line
from models.mixins.has_order_index import HasOrderIndex
from models.mixins.has_slug import HasSlug
from models.mixins.is_closable import IsClosable
from models.mixins.is_searchable import IsSearchable
from models.mixins.is_secret import IsSecret
from models.sector import Sector
from util.entity_count_cache import get_cached_ascent_count, get_cached_line_count
from util.secret_service import SecretService
from util.topo_tab_counts import count_gallery_images, count_root_comments


class Crag(HasSlug, HasOrderIndex, IsSearchable, IsClosable, IsSecret, BaseEntity):
    """
    Model of a climbing crag. Could be e.g. "Glees". Contains one or more sectors.
    """

    __tablename__ = "crags"

    searchable_type = SearchableItemTypeEnum.CRAG
    name = db.Column(db.String(120), nullable=False)
    short_description = db.Column(db.Text, nullable=True)
    description = db.Column(db.Text, nullable=True)
    rules = db.Column(db.Text, nullable=True)
    portrait_image_id = db.Column(UUID(), db.ForeignKey("files.id"), nullable=True)
    portrait_image = db.relationship("File", lazy="joined")
    sectors = db.relationship(
        "Sector", cascade="all,delete", backref="crag", lazy="select", order_by="Sector.order_index.asc()"
    )
    rankings = db.relationship("Ranking", cascade="all,delete", lazy="select")
    map_markers = db.relationship("MapMarker", back_populates="crag")
    default_boulder_scale = db.Column(db.String(32), nullable=True)
    default_sport_scale = db.Column(db.String(32), nullable=True)
    default_trad_scale = db.Column(db.String(32), nullable=True)
    blocweather_url = db.Column(db.String(255), nullable=True)

    @hybrid_property
    def line_count(self):
        cached = get_cached_line_count(self)
        if cached is not None:
            return cached
        query = (
            db.session.query(func.count(Line.id))
            .join(Area, Line.area_id == Area.id)
            .join(Sector, Area.sector_id == Sector.id)
            .where(Sector.crag_id == self.id)
        )
        query = SecretService.apply_line_filter(query)
        return query.scalar()

    @hybrid_property
    def ascent_count(self):
        cached = get_cached_ascent_count(self)
        if cached is not None:
            return cached
        query = (
            db.session.query(func.count(Ascent.id))
            .join(Line)
            .join(Area, Line.area_id == Area.id)
            .join(Sector, Area.sector_id == Sector.id)
            .where(Sector.crag_id == self.id)
        )
        query = SecretService.apply_line_filter(query)
        return query.scalar()

    @hybrid_property
    def comment_count(self):
        return count_root_comments("Crag", self.id)

    @hybrid_property
    def image_count(self):
        return count_gallery_images("Crag", self.id)

    @classmethod
    def find_max_order_index(cls) -> int:
        max_order_index = db.session.query(func.max(cls.order_index)).first()

        if len(max_order_index) == 0 or max_order_index[0] is None:
            return -1

        return max_order_index[0]

File: client/src/app/models/crag.ts
import { AbstractModel } from './abstract-model';
import { File } from './file';
import { Sector } from './sector';
import { MapMarker } from './map-marker';
import {
  deserializeClosableAttributes,
  IsClosable,
  serializeClosableAttributes,
} from './mixins/is-closable';
import {
  deserializeOrderIndexAttributes,
  HasOrderIndex,
} from './mixins/has-order-index';
import { deserializeSlugAttributes, HasSlug } from './mixins/has-slug';
import { topoCragRouterLink } from './topo-router-link';

/**
 * Model of a climbing crag.
 */
export class Crag extends IsClosable(HasOrderIndex(HasSlug(AbstractModel))) {
  name: string;
  description: string;
  shortDescription: string;
  rules: string;
  portraitImage: File;
  sectors: Sector[];
  lineCount: number;
  ascentCount: number;
  routerLink: string | null;
  secret: boolean;
  mapMarkers: MapMarker[];
  defaultBoulderScale: string | null = null;
  defaultSportScale: string | null = null;
  defaultTradScale: string | null = null;
  blocweatherUrl: string | null = null;

  /**
   * Parses a crag.
   *
   * @param payload Crag json payload.
   * @return Parsed Crag.
   */
  public static deserialize(payload: any): Crag {
    const crag = new Crag();
    AbstractModel.deserializeAbstractAttributes(crag, payload);
    deserializeClosableAttributes(crag, payload);
    deserializeSlugAttributes(crag, payload);
    deserializeOrderIndexAttributes(crag, payload);
    crag.name = payload.name;
    crag.description = payload.description;
    crag.shortDescription = payload.shortDescription;
    crag.rules = payload.rules;
    crag.secret = payload.secret;
    crag.portraitImage = payload.portraitImage
      ? File.deserialize(payload.portraitImage)
      : null;
    crag.sectors = payload.sectors
      ? payload.sectors.map(Sector.deserialize)
      : null;
    crag.lineCount = payload.lineCount;
    crag.ascentCount = payload.ascentCount;
    crag.mapMarkers = payload.mapMarkers
      ? payload.mapMarkers.map(MapMarker.deserialize)
      : null;
    crag.routerLink = topoCragRouterLink(crag);
    crag.defaultBoulderScale = payload.defaultBoulderScale;
    crag.defaultSportScale = payload.defaultSportScale;
    crag.defaultTradScale = payload.defaultTradScale;
    crag.blocweatherUrl = payload.blocweatherUrl ?? null;
    return crag;
  }

  /**
   * Marshals a Crag.
   *
   * @param crag Crag to marshall.
   * @return Marshalled Crag.
   */
  public static serialize(crag: Crag): any {
    return {
      ...serializeClosableAttributes(crag),
      ...{
        name: crag.name,
        description: crag.description,
        shortDescription: crag.shortDescription,
        rules: crag.rules,
        secret: crag.secret,
        portraitImage: crag.portraitImage ? crag.portraitImage.id : null,
        mapMarkers: crag.mapMarkers
          ? crag.mapMarkers.map(MapMarker.serialize)
          : null,
        defaultBoulderScale: crag.defaultBoulderScale,
        defaultSportScale: crag.defaultSportScale,
        defaultTradScale: crag.defaultTradScale,
        blocweatherUrl: crag.blocweatherUrl,
      },
    };
  }
}
